### PR TITLE
Adds tethys-platform

### PR DESCRIPTION
**Pull request by swainn**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by swainn** [comment]
I agree to be a maintainer

**Comment by sdc50** [comment]
I agree to be a maintainer

**Comment by swainn** [comment]
@conda-forge/help-python this is good to go.

**Comment by swainn** [comment]
> @conda-forge/help-python this is good to go.

@conda-forge/help-python any feedback on this yet?

File: recipes/tethys-platform/bld.bat [hunk 1]
```diff
@@ -0,0 +1,2 @@
+set PBR_VERSION=3.4.3
```
Comment (carterbox) [inline recipes/tethys-platform/bld.bat:1]:
```suggestion
set PBR_VERSION=$PKG_VERSION
```
https://docs.conda.io/projects/conda-build/en/latest/user-guide/environment-variables.html#environment-variables-set-during-the-build-process

File: recipes/tethys-platform/build.sh [hunk 2]
```diff
@@ -0,0 +1 @@
+PBR_VERSION=3.4.3 $PYTHON -m pip install . -vv
```
Comment (carterbox) [inline recipes/tethys-platform/build.sh:1]:
```suggestion
PBR_VERSION=%PKG_VERSION% $PYTHON -m pip install . -vv
```
https://docs.conda.io/projects/conda-build/en/latest/user-guide/environment-variables.html#environment-variables-set-during-the-build-process

File: recipes/tethys-platform/meta.yaml [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+    - tethys_cli
+    - tethys_compute
+    - tethys_config
+    - tethys_gizmos
+    - tethys_portal
+    - tethys_quotas
+    - tethys_sdk
+    - tethys_services
+  commands:
+    - tethys version
```
Comment (carterbox) [inline recipes/tethys-platform/meta.yaml:91]:
```suggestion
test:
  requires:
    - pip
  imports:
    - tethys_apps
    - tethys_cli
    - tethys_compute
    - tethys_config
    - tethys_gizmos
    - tethys_portal
    - tethys_quotas
    - tethys_sdk
    - tethys_services
  commands:
    - pip check
    - tethys version
```

File: recipes/tethys-platform/meta.yaml [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+    - django-simple-captcha =0.5.*
+    - django-recaptcha2 =1.4.*
+    - django-mfa2 <2.5
+    - social-auth-app-django =4.0.*
+    - requests-mock =1.9.*
+    - factory_boy =3.2.*
+    - pillow =9.2.*
+    - pip =22.1.*
+    - future =0.18.*
+    - pbr =5.9.*
```
Comment (carterbox) [inline recipes/tethys-platform/meta.yaml:75]: Why are pip and pbr required but also ignored from run_exports? Why are they required at runtime?

File: recipes/tethys-platform/meta.yaml [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+    - git
+    - conda
```
Comment (carterbox) [inline recipes/tethys-platform/meta.yaml:77]: Why are git and conda required at runtime?

**Review by carterbox — CHANGES_REQUESTED** [review]

File: recipes/tethys-platform/meta.yaml (hunk 5 above)
Comment (swainn) [reply recipes/tethys-platform/meta.yaml:77]: pip and conda are required because Tethys uses their APIs in the `tethys install` command to install dependencies of apps that it is installing.

File: recipes/tethys-platform/meta.yaml (hunk 4 above)
Comment (swainn) [reply recipes/tethys-platform/meta.yaml:75]: pip and conda are required because Tethys uses their APIs in the `tethys install` command to install dependencies of apps that it is installing.

File: recipes/tethys-platform/bld.bat (hunk 1 above)
Comment (swainn) [reply recipes/tethys-platform/bld.bat]: Done.

File: recipes/tethys-platform/meta.yaml (hunk 3 above)
Comment (swainn) [reply recipes/tethys-platform/meta.yaml]: Done.

File: recipes/tethys-platform/meta.yaml (hunk 4 above)
Comment (swainn) [reply recipes/tethys-platform/meta.yaml:75]: I'm not sure I understand your comment about them being ignored from run_exports. Doc link?

File: recipes/tethys-platform/meta.yaml (hunk 5 above)
Comment (swainn) [reply recipes/tethys-platform/meta.yaml]: Removed git

**Comment by swainn** [comment]
@conda-forge/help-python and @carterbox I've resolved your comments to the best of my understanding. Please help me to better under stand the comment about ignoring them from run_exports. Thank you.

File: recipes/tethys-platform/meta.yaml (hunk 4 above)
Comment (carterbox) [reply recipes/tethys-platform/meta.yaml]:
```yaml
  ignore_run_exports:
    - pip
    - pbr
```
Why does your recipe include this section?

https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#export-runtime-requirements

Comment (carterbox) [reply recipes/tethys-platform/meta.yaml]: run_exports are packages which are implicitly added to the run requirements of your recipe when you explicity add a related package to your host/build requirements section. It is used for compiled libraries to ensure that linked shared libraries are included. Why do your recipe want to prevent `pip` and `pbr` from being exported in this manner? I'm not even sure that any packages would export `pip` or `pbr`.

Comment (carterbox) [reply recipes/tethys-platform/meta.yaml]: > Removed pbr.

Looks like pbr has a module which is also imported at runtime.

https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=559916&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=2664

Comment (swainn) [reply recipes/tethys-platform/meta.yaml]: Tests failed when pbr was removed as runtime dependency. It is required for the version command.

See: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=559916&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823

Comment (swainn) [reply recipes/tethys-platform/meta.yaml]: > ```yaml
>   ignore_run_exports:
>     - pip
>     - pbr
> ```
> 
> Why does your recipe include this section?
> 
> https://docs.conda.io/projects/conda-build/en/latest/resources/define-metadata.html#export-runtime-requirements

Ah yes. That was added due to this error on the Windows build check: https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=546753&view=logs&j=240f1fee-52bc-5498-a14a-8361bde76ba0&t=2fb5b0a7-737c-5f5c-8f3f-f6db6174bacf&l=2151

Comment (carterbox) [reply recipes/tethys-platform/meta.yaml]: > Ah yes. That was added due to this error on the Windows build check

The linked error message is as follows:

> [noarch_python] Noarch package contains binary script: tethys.exe

This error says that conda found a platform specific binary in a noarch package. No arch means it runs on any platform. A windows binary cannot run on Linux. This should be unrelated to whether pip or pbr are in the requirements/run section, and prevents a package from being valid noarch. Is your package [noarch or not?](https://conda-forge.org/docs/maintainer/knowledge_base.html#noarch-python)

Also, we do not allow pre-compiled binaries. You must use our compiler toolchain to compile binaries from source.

Comment (dopplershift) [reply recipes/tethys-platform/meta.yaml]: Specifically, I think it's due to the script listed `pyproject.toml`. Those are [not allowed for noarch](https://conda-forge.org/docs/maintainer/knowledge_base.html?highlight=noarch#noarch-python). If you converted that to a proper `console_script` entrypoint, that would allow setuptools to handle creating the proper executable on both Windows and *nix. I think most of the effort would be in properly specifying the metadata.

Comment (BastianZim) [reply recipes/tethys-platform/meta.yaml]: Poetry scripts are converted to entry points according to https://github.com/python-poetry/poetry/issues/927
They just need to be explicitly listed as entry points in the meta.yaml. Was that the case when the error occurred?

Comment (carterbox) [reply recipes/tethys-platform/meta.yaml]: Maybe your console_scipts entry point is configured incorrectly?

Comment (dopplershift) [reply recipes/tethys-platform/meta.yaml]: I don't see poetry being used here? So I think the problem is that it's declared as a setuptools "script", while it needs to be a "console_script".

Comment (BastianZim) [reply recipes/tethys-platform/meta.yaml]: Ahh true you're right :)

Comment (carterbox) [reply recipes/tethys-platform/meta.yaml]: I think that error message is from an old commit before the entry point was registered with the recipe. I expect it would not occur now.

File: recipes/tethys-platform/meta.yaml [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+source:
+  url: https://github.com/tethysplatform/tethys/archive/refs/tags/{{ version }}.tar.gz
+  sha256: 3775d5599e756206204e09ca9a2114c8866fad7fd29999b256343c847ada3b1f
+
+build:
+  noarch: python
+  number: 0
+  ignore_run_exports:
+    - pip
+    - pbr
```
Comment (carterbox) [inline recipes/tethys-platform/meta.yaml:17]:
```suggestion
```
I believe this is unrelated to failed Windows builds from earlier. Regardless, it shouldn't be necessary.

**Review by carterbox — CHANGES_REQUESTED** [review]

File: recipes/tethys-platform/meta.yaml (hunk 4 above)
Comment (carterbox) [reply recipes/tethys-platform/meta.yaml]: @BastianZim @dopplershift, the "console_script" syntax is removed with the shift to `pyproject.toml`

https://setuptools.pypa.io/en/latest/userguide/quickstart.html?highlight=console_scripts#entry-points-and-automatic-script-creation

Comment (swainn) [reply recipes/tethys-platform/meta.yaml]: >This error says that conda found a platform specific binary in a noarch package. No arch means it runs on any platform. A windows binary cannot run on Linux. This should be unrelated to whether pip or pbr are in the requirements/run section, and prevents a package from being valid noarch. Is your package [noarch or not?](https://conda-forge.org/docs/maintainer/knowledge_base.html#noarch-python)

As far as I can tell, the project meets the requirements listed in that link for noarch. We are not vendoring any pre-compiled binaries. `tethys.exe` is generated by setuptools as the entrypoint for our console script, which is specified in the setup.cfg (see below).

>Specifically, I think it's due to the script listed pyproject.toml. Those are [not allowed for noarch](https://conda-forge.org/docs/maintainer/knowledge_base.html?highlight=noarch#noarch-python). 

Tethys Platform 3.X doesn't have a pyproject.toml (you probably looked at the master branch)? Tethys 3.4 is on the tethys_3.4 branch.

> If you converted that to a proper console_script entrypoint, that would allow setuptools to handle creating the proper executable on both Windows and *nix. I think most of the effort would be in properly specifying the metadata.

> Maybe your console_scipts entry point is configured incorrectly?

 The console script is configured in the [setup.cfg](https://github.com/tethysplatform/tethys/blob/release-3.4/setup.cfg#L34) as follows:

```ini
[entry_points]
console_scripts =
    tethys = tethys_cli:tethys_command
```

How would this need to be changed to make it a "proper" `console_script`?

> Poetry scripts are converted to entry points according to https://github.com/python-poetry/poetry/issues/927
They just need to be explicitly listed as entry points in the meta.yaml. Was that the case when the error occurred?

We are not using poetry as our packaging / build system. We use setuptools.

Comment (dopplershift) [reply recipes/tethys-platform/meta.yaml]: Ah I missed that, but it looks like the missing entry point was, as suggested by @carterbox, just outdated.

Comment (swainn) [reply recipes/tethys-platform/meta.yaml]: I think I know what happened now. I added `pip` and `pbr` to the `ignore_run_exports` section thinking that the warnings were related to the `tethys.exe` error. But I also added the `entry_points` section to the recipe in the same commit thinking both changes were necessary and the build passed.

I've now removed the `entry_points` section as @carterbox suggested and the Windows build still works because I've specified the `entry_point`. Sorry for the confusion.
